core/app/models/spree/payment/processing.rb

```diff
@@ -1,11 +1,36 @@
 module Spree
   class Payment < Spree::Base
     module Processing
+      # "process!" means:
+      #   - Do nothing when:
+      #     - There is no payment method
+      #     - The payment method does not require a source
+      #     - The payment is in the "processing" state
+      #     - 'auto_capture?' is false and the payment is already authorized.
+      #   - Raise an exception when:
+      #     - The source is missing or invalid
+      #     - The payment is in a state that cannot transition to 'processing'
+      #       (failed/void/invalid states). Note: 'completed' can transition to
+      #       'processing' and thus calling #process! on a completed Payment
+      #       will attempt to re-authorize/re-purchase the payment.
+      #   - Otherwise:
+      #     - If 'auto_capture?' is true:
+      #       - Call #purchase on the payment gateway. (i.e. authorize+capture)
+      #         even if the payment is already failed/completed/etc.
+      #     - Else:
+      #       - Call #authorize on the payment gateway even if the payment is
+      #         already authorized/failed/completed/etc.
       def process!
-        if payment_method && payment_method.auto_capture?
+        return if payment_method.nil?
+
+        if payment_method.auto_capture?
           purchase!
         else
-          authorize!
+          if pending?
```

Comment:
Should this loop be moved up a level?

If auto capture is on, and this payment is pending, right now we will attempt to `purchase!` which I don't think is correct.